compiler/damlc/daml-doc/src/DA/Daml/Doc/Driver.hs

```diff
@@ -18,67 +21,99 @@ import Development.IDE.Types.Location
+    , do_docTitle :: Maybe T.Text
+    , do_combine :: Bool
+    }
 
 data InputFormat = InputJson | InputDaml
     deriving (Eq, Show, Read)
 
-damlDocDriver :: InputFormat
-              -> IdeOptions
-              -> FilePath
-              -> DocFormat
-              -> Maybe FilePath
-              -> [DocOption]
-              -> [NormalizedFilePath]
-              -> IO ()
-damlDocDriver cInputFormat ideOpts output cFormat prefixFile options files = do
+data OutputFormat = OutputJson | OutputHoogle | OutputDocs RenderFormat
+    deriving (Eq, Show, Read)
+
+-- | Run damldocs!
+damlDocDriver :: DamldocOptions -> IO ()
+damlDocDriver options = do
+    docData <- inputDocData options
+    renderDocData options (transformDocData options docData)
+
+-- | Load doc data, either via the DAML typechecker or via JSON files.
+inputDocData :: DamldocOptions -> IO [ModuleDoc]
+inputDocData DamldocOptions{..} = do
 
     let printAndExit errMsg = do
-            putStrLn $ "Error processing input from " <> unwords (map fromNormalizedFilePath files) <> "\n"
-                     <> errMsg
-            fail "Aborted."
+            hPutStr stderr . unlines $
+                [ unwords
+                    $ "Error processing input from"
+                    : map fromNormalizedFilePath do_inputFiles
+                , errMsg
+                ]
+            exitFailure
 
-    let onErrorExit act =
-            act >>= either (printAndExit . renderDiags) pure
         renderDiags = T.unpack . showDiagnosticsColored
+        onErrorExit act = act >>= either (printAndExit . renderDiags) pure
+
+    case do_inputFormat of
+        InputJson -> do
+            input <- mapM (BS.readFile . fromNormalizedFilePath) do_inputFiles
+            let mbData = map (AE.eitherDecode . LBS.fromStrict) input
+            applyAnnotations <$> concatMapM (either printAndExit pure) mbData
 
-    docData <- case cInputFormat of
-            InputJson -> do
-                input <- mapM (BS.readFile . fromNormalizedFilePath) files
-                let mbData = map AE.eitherDecode input :: [Either String [ModuleDoc]]
-                applyAnnotations <$> concatMapM (either printAndExit pure) mbData
+        InputDaml -> onErrorExit . runExceptT $
+            fmap (applyTransform do_docOptions)
+                (mkDocs do_ideOptions do_inputFiles)
 
-            InputDaml ->
-                onErrorExit $ runExceptT
-                            $ fmap (applyTransform options)
-                            $ mkDocs ideOpts files
+-- | Transform doc data, applying annotations and
+transformDocData :: DamldocOptions -> [ModuleDoc] -> [ModuleDoc]
+transformDocData DamldocOptions {..} = applyTransform do_docOptions
 
-    prefix <- maybe (pure "") BS.readFile prefixFile
+-- | Output doc data.
+renderDocData :: DamldocOptions -> [ModuleDoc] -> IO ()
+renderDocData DamldocOptions{..} docData = do
+    templateM <- mapM (fmap T.decodeUtf8 . BS.readFile) do_docTemplate
```

Comment:
```suggestion
    templateM <- mapM T.readFileUtf8 do_docTemplate
```
with the helper from `Data.Text.Extended`.

Reply:
thanks!